README.md
<!-- badges: start -->

[![CRAN_version](http://www.r-pkg.org/badges/version/atpolR)](https://cran.r-project.org/package=atpolR) [![R-CMD-check](https://github.com/gsapijaszko/atpolR/actions/workflows/R-CMD-check.yaml/badge.svg)](https://github.com/gsapijaszko/atpolR/actions/workflows/R-CMD-check.yaml) ![month](http://cranlogs.r-pkg.org/badges/atpolR) ![total](http://cranlogs.r-pkg.org/badges/grand-total/atpolR)

<!-- badges: end -->

# atpolR

ATPOL grid implementation in R


NEWS.md
# atpolR: 0.1.2
## News
  * changed classification of background raster
  * added options to plot water layer (`water = TRUE`) 
  * added possibility to plot b-w map with `colors = FALSE` and `water = FALSE`
  * removed `png` plot bug

# atpolR: 0.1.1
## News
  * Updated `plot_points_on_atpol()` and vignette.

# atpolR: 0.1.0
## News
  * First release


R/atpol_grids.R
utils::globalVariables(c("Name", ".data"))
#
#' .atpol10k_full reads ATPOL 10km x 10km grid from file and returns it as sf object
#' @importFrom terra vect
#' @importFrom sf st_as_sf
#' @return Simple Feature (sf) grid of polygons for 10km x 10km ATPOL grid
#' @usage .atpol10k_full()
#' @noRd
#'
.atpol10k_full <- function() {
  if(file.exists(system.file("extdata", "atpol10k.Rds", package = "atpolR"))) {
    v <- terra::vect(readRDS(system.file("extdata", "atpol10k.Rds", package = "atpolR", mustWork = TRUE)))
    v <- sf::st_as_sf(v)
    return(v)
  }
  else {
    warning("No extdata/atpol10k.Rds data found.")
  }
}

#' atpol10k returns ATPOL grid 10x10 km and returns it as sf object
#' @importFrom sf st_centroid
#' @return Simple Feature (sf) grid of polygons for 10km x 10km ATPOL grid
#' @export
#' @usage atpol10k()
#'
atpol10k <- function() {
  grid10k <- subset(.atpol10k_full(), !grepl("AA|AF|AG|BG|CG", .atpol10k_full()$Name))
  grid10k$centroid <- sf::st_centroid(grid10k$geometry)
  return(grid10k)
}

#' atpol100k creates ATPOL grid 100km x 100km and returns it as sf object
#' @importFrom stats aggregate setNames
#' @return Simple Feature (sf) grid of polygons for 100km x 100km ATPOL grid
#' @export
#' @usage atpol100k()
#'
atpol100k <- function() {
  grid100k <- setNames(aggregate(atpol10k()[,"geometry"], by = list(substr(atpol10k()$Name, 1, 2)), mean), c("Name", "geometry"))
  return(grid100k)
}

#' atpol1k creates ATPOL grid 1km x 1km and returns it as sf object
#' @importFrom terra vect project
#' @importFrom sf st_as_sf
#' @return Simple Feature (sf) grid of polygons for 1km x 1km ATPOL grid
#' @param grid any valid ATPOL 10km grid like "BE23" or "DC58"
#' @export
#' @examples
#' atpol1k("BE23")
#'
atpol1k <- function(grid) {
  g <- toupper({{grid}})
  if (all(g[] %in% atpol10k()$Name) == FALSE) {
    stop("Please make sure all grids are proper ATPOL 10km grid squares, like 'BE23'")
  }
  else {
    v <- terra::vect()
    for(gr in g) {
      for(x1 in c(0:9)) {
        message(paste0("Generating grid for ",gr,x1,"..."))
        for(y1 in c(0:9)) {
          g0 = structure(c(
            grid_to_latlon(paste0(gr,x1,y1), 0, 0)[2],grid_to_latlon(paste0(gr,x1,y1), 0, 1)[2],
            grid_to_latlon(paste0(gr,x1,y1), 1, 1)[2],grid_to_latlon(paste0(gr,x1,y1), 1, 0)[2],
            grid_to_latlon(paste0(gr,x1,y1), 0, 0)[1],grid_to_latlon(paste0(gr,x1,y1), 0, 1)[1],
            grid_to_latlon(paste0(gr,x1,y1), 1, 1)[1],grid_to_latlon(paste0(gr,x1,y1), 1, 0)[1]),
            .Dim = c(4L, 2L))
          a <- terra::vect(g0, crs = "EPSG:4326", "polygons", atts = data.frame("Name" = paste0(gr,x1,y1))) |>
            terra::project("EPSG:2180")
          if (dim(v)[1] == 0) {
            v <- a
          } else { v <- v+a}
        }
      }
    }
    v <- terra::project(v, "EPSG:2180") |>
      sf::st_as_sf()
    return(v)
  }
}

#' atpol_div creates ATPOL grid divided by 2, 4 or 5 (based on divider parameter) and returns it as sf object. Useful for grids like 5 x 5 km (divider = 2), 250 x 250 m (divider = 4) or 20 x 20 m (divider = 5). For details see \insertCite{vereyStandaryzacjaZapisuPodzialow2018;textual}{atpolR}
#' @importFrom terra vect project
#' @importFrom sf st_as_sf
#' @return Simple Feature (sf) grid of polygons for ATPOL grid divided by 2, 4 or 5
#' @param grid any valid ATPOL grid like "BE" or "DC5128"
#' @param divider divide by parameter: 2, 4, 5
#' @export
#' @usage atpol_div(grid, divider)
#' @examples
#' \donttest{
#' atpol_div("BE", 2)
#' atpol_div(grid = c("BE23", "DC5128"), divider = 4)
#' }
#' @references
#'     \insertAllCited{}
#'
#'
atpol_div <- function(grid, divider) {
  g <- toupper({{grid}})
  if(!divider %in% c(2,4,5)) {
    stop("Please make sure divider is one of: 2, 4, 5")
  }
  if (all(substr(g, 1, 2)[] %in% unique(substr(atpol10k()$Name, 1,2))) == FALSE) {
    stop("Please make sure all grids are proper ATPOL grid squares, like 'BE' or 'CE2345'")
  } else if (all(nchar(g[]) %% 2 == 0) == FALSE) {
    stop("Please make sure length of all grid fields is 2, 4, 6, 8, 10 or 12")
  } else {
    v <- terra::vect()

    if(divider == 2 ) {subgrid <- "d"}
    else if (divider == 4 ) {subgrid <- "c"}
    else if (divider == 5 ) {subgrid <- "p"}

    for(gr in g) {
      for(x1 in c(0:(divider-1))) {
        message(paste0("Generating grid for ",gr,x1,"..."))
        for(y1 in c(0:(divider-1))) {
          g0 = structure(c(
            grid_to_latlon(gr, xoffset = 0+y1/divider, yoffset = 0+x1/divider)[2],grid_to_latlon(gr, xoffset = (1+y1)/divider, yoffset = 0+x1/divider)[2],
            grid_to_latlon(gr, xoffset = (1+y1)/divider, yoffset = (1+x1)/divider)[2],grid_to_latlon(gr, xoffset = 0+y1/divider, yoffset = (1+x1)/divider)[2],
            grid_to_latlon(gr, xoffset = 0+y1/divider, yoffset = 0+x1/divider)[1],grid_to_latlon(gr, xoffset = (1+y1)/divider, yoffset = 0+x1/divider)[1],
            grid_to_latlon(gr, xoffset = (1+y1)/divider, yoffset = (1+x1)/divider)[1],grid_to_latlon(gr, xoffset = 0+y1/divider, yoffset = (1+x1)/divider)[1]),
            .Dim = c(4L, 2L))
          a <- terra::vect(g0, crs = "EPSG:4326", "polygons", atts = data.frame("Name" = paste0(gr,subgrid,x1,y1))) |>
            terra::project("EPSG:2180")
          if (dim(v)[1] == 0) {
            v <- a
          } else { v <- v+a }
        }
      }
    }
    v <- terra::project(v, "EPSG:2180") |>
      sf::st_as_sf()
    return(v)
  }
}


R/miscRscript.r
# preparation of basic 10km x 10km grid used in other functions and saving it as data file
# the data is in terra::vect format
if(!file.exists("inst/extdata/atpol10k.Rds")) {
  v <- terra::vect()
  for (X in LETTERS[1:7]) {
    for (Y in LETTERS[1:7]) {
      message(paste0("generating grids for ", X,Y))
      for(x1 in c(0:9)) {
        for(y1 in c(0:9)) {
          g0 = structure(c(
            grid_to_latlon(paste0(X,Y,x1,y1), 0, 0)[2],grid_to_latlon(paste0(X,Y,x1,y1), 0, 1)[2],
            grid_to_latlon(paste0(X,Y,x1,y1), 1, 1)[2],grid_to_latlon(paste0(X,Y,x1,y1), 1, 0)[2],
            grid_to_latlon(paste0(X,Y,x1,y1), 0, 0)[1],grid_to_latlon(paste0(X,Y,x1,y1), 0, 1)[1],
            grid_to_latlon(paste0(X,Y,x1,y1), 1, 1)[1],grid_to_latlon(paste0(X,Y,x1,y1), 1, 0)[1]),
            .Dim = c(4L, 2L))
          a <- terra::vect(g0, crs = "EPSG:4326", "polygons", atts = data.frame("Name" = paste0(X,Y,x1,y1))) |>
            terra::project("EPSG:2180")

          if (dim(v)[1] == 0) {
            v <- a
          } else { v <- v+a}
        }
      }
    }
  }
  v <- terra::project(v, "EPSG:2180")
  terra::saveRDS(v, file = "inst/extdata/atpol10k.Rds")
}

# preparing the simplified boundary data of Poland
if(!file.exists("inst/extdata/pl_boundary.Rds")) {
  # getting PL boundary data
  boundary <- geodata::gadm(country = "POL", level = 1, path = tempdir(), version = "3.6")
  boundary <- sf::st_as_sf(boundary)
  sf::st_crs(boundary) <- 4326
  boundary <- sf::st_union(boundary) |>
    sf::st_cast(to = "MULTILINESTRING") |>
    sf::st_as_sf() |>
    sf::st_transform(crs = "EPSG:2180") |>
    sf::st_simplify(dTolerance = 100)
  base::saveRDS(boundary, file = "inst/extdata/pl_boundary.Rds")
}


if(!file.exists("inst/extdata/cr.tif")) {
  # preparing elevation raster for Poland with BBox of ATPOL
  # CRS = EPSG:2180
  countries <- c("POL", "CZE", "SVK", "UKR", "BLR", "LTU")
  for (cntry in countries) {
    geodata::elevation_30s(country = cntry, path = tempdir(), mask = FALSE)
  }
  bbox <- sf::st_bbox(atpolR::atpol10k())
  ext <- terra::ext(c(bbox[1], bbox[3], bbox[2], bbox[4]))

  r <- list.files(path = tempdir(), pattern = ".+._elv.tif", full.names = TRUE)
  r <- lapply(r, terra::rast)
  r <- do.call(terra::merge, r)
  terra::crs(r) <- "EPSG:4326"
  r<-terra::project(r, "EPSG:2180")
  r <- terra::crop(r, ext)
  r[r<=0] <- 10
  r <- terra::classify(r, c(0,50,100,200,300,400,500,600, 700, 800, 1000, 1200, 1500, 1700, 2000, 2500), brackets = FALSE, others = 1)
  terra::writeRaster(r, filename = "inst/extdata/cr.tif", overwrite = TRUE)
}


R/latlon_to_grid.R
# All functions based on https://sourceforge.net/p/atpol/code/HEAD/tree/
# 19E, 52N

.A52 = 0.907571211037051
.RAD_TO_DEG <- 57.2957795130823
.EARTH_RADIUS <- 6390
.LAMBDA_ZERO <- 19
.X_ZERO <- 330
.Y_ZERO <- 350

.latlon_to_xy <- function(lat, lon) {
  lon = lon - .LAMBDA_ZERO
  lon = lon / .RAD_TO_DEG
  lat = lat / .RAD_TO_DEG
  r = cos(.A52) / sin(.A52) - tan(lat - .A52)
  x = r * sin(lon * sin(.A52))
  y = r * cos(lon * sin(.A52))
  y = y - cos(.A52) / sin(.A52)
  x = x * .EARTH_RADIUS + .X_ZERO
  y = y * .EARTH_RADIUS + .Y_ZERO
  a <- c(x,y)
  return(a)
}

.xy_to_latlon <- function(x, y) {
  x = (x - .X_ZERO) / .EARTH_RADIUS
  y = (y - .Y_ZERO) / .EARTH_RADIUS
  y = y + cos(.A52) / sin(.A52)
  lon = atan(x / y) / sin(.A52)
  r = sqrt(x * x + y * y)
  lat = .A52 - atan(r - cos(.A52) / sin(.A52))
  lat = lat * .RAD_TO_DEG
  lon = lon * .RAD_TO_DEG + .LAMBDA_ZERO
  a <- c(lat, lon)
  return(a)
}

.xy_to_grid <- function (x, y, len) {
  grid = ""
  xoffset = 0;
  yoffset = 0;
  if ((x >= 0) & (x < 700) & (y >= 0) & (y < 700) & (len >= 2) & (len <= 12) & ((len %% 2) == 0)) {
    xs = round(x * 1000);
    ys = round(y * 1000);
    xs = as.character(xs);
    ys = as.character(ys);
    xs = stringr::str_pad(xs, 6, '0', side = "left");
    ys = stringr::str_pad(ys, 6, '0', side = "left");
    grid = paste0(intToUtf8(utf8ToInt(substr(xs,1,1))+17),intToUtf8(utf8ToInt(substr(ys,1,1))+17))
    for (i in 2:6) {
      grid = paste0(grid, substr(ys,i,i),substr(xs,i,i));
    }
    xoffset = as.character(as.integer(round(x * 1000000)))
    yoffset = as.character(as.integer(round(y * 1000000)))
    xoffset = stringr::str_pad(xoffset, 9, '0', side = "left");
    yoffset = stringr::str_pad(yoffset, 9, '0', side = "left");
    xoffset = paste0("0.", substr(xoffset, -9 + len / 2, 9))
    yoffset = paste0("0.", substr(yoffset, -9 + len / 2, 9))
    xoffset = as.double(xoffset);
    yoffset = as.double(yoffset);
  }
  a <- c(substr(grid, 0, len), xoffset, yoffset)
  return(a)
}

.grid_to_xy <- function(grid, xoffset, yoffset) {
  grid = toupper(grid);
  l = nchar(grid);
  if (l > 0 & grepl('^[A-G][A-G][0-9]*',grid)) {
    xs = intToUtf8(utf8ToInt(substr(grid,1,1))-17);
    ys = intToUtf8(utf8ToInt(substr(grid,2,2))-17);
    if (l/2 > 1){
      for (i in 2:(l/2)) {
        xs = paste0(xs, substr(grid, 2*i, 2*i))
        ys = paste0(ys, substr(grid, 2*i-1, 2*i-1))
      }
    }
    if(l/2 < 6) {
      for (i  in ((l/2)+1):6) {
        xs = paste0(xs, "0")
        ys = paste0(ys, "0")
      }
    }
    x <- as.numeric(xs)/1000 + 10^(3-l/2) * as.numeric(xoffset);
    y <- as.numeric(ys)/1000 + 10^(3-l/2) * as.numeric(yoffset);
  }
  a <- c(x, y)
  return(a);
}

#' grid_to_latlon converts the ATPOL grid to latitude and longitude.
#' With \code{xoffset = 0} and \code{yoffset = 0} parameters it returns coordinates of the upper left corner of the grid.
#' @param grid An ATPOL grid, ex. "GF2345".
#' @param xoffset An offset in X, where 0 is for left, and 1 for right side of the grid. The default value is 0.5, which corresponds to middle of the grid.
#' @param yoffset An offset in Y, where 0 is for top, and 1 for bottom side of the grid. The default value is 0.5, which corresponds to middle of the grid.
#' @return latitude and longitude of ATPOL grid (default centroid) as pair of numerics
#' @examples
#' grid_to_latlon("BE21")
#' grid_to_latlon("BE21", 0, 0)
#' @references \url{https://atpol.sourceforge.io/}
#' @export
#'
grid_to_latlon <- function(grid, xoffset=0.5, yoffset=0.5) {
  xy = .grid_to_xy(grid, xoffset, yoffset)
  latlon = .xy_to_latlon(xy[1], xy[2]);
  return(latlon)
}

#' latlon_to_grid(lat, lon, length) converts geographical coordinates to ATPOL grid of given length
#' @import stringr
#' @param lat Latitude in degrees, ex. 51.123456
#' @param lon Longitude in degrees, ex. 17.234567
#' @param length Desired ATPOL grid length, which can be 2, 4, 6, 8, 10 or 12,
#' @return grid, ex. BE, BE23, BE2357, etc.
#' @examples latlon_to_grid(51, 17, 2)
#' @examples latlon_to_grid(51, 17, 6)
#' @references \url{https://atpol.sourceforge.io/}
#' @export
latlon_to_grid <- function(lat, lon, length) {
  xy = .latlon_to_xy(lat, lon);
  grid = .xy_to_grid(xy[1], xy[2], length);
  return(grid[1])
}

# a<- latlon_to_grid(50,16, 4)
# grid_to_latlon("BF61", 0, 0)


R/check_atpol_square.R
#' Reverse engineering of published ATPOL grids
#'
#' check_atpol_square() do a reverse engineering of published ATPOL grids species,
#' especially those published in \insertCite{zajacAtlasRozmieszczeniaRoslin2001;textual}{atpolR}
#' @importFrom sf st_buffer
#' @importFrom terra crop vect
#' @importFrom Rdpack reprompt
#' @param centroid Simple Feature point geometry for which the check is performed,
#' usually it corresponds to centroid of ATPOL 10km x 10km grid
#' @param raster geocoded raster, it has to be in EPSG:2180 projection
#' @param distance \code{st_buffer} distance from centroid point for which the check is done, default 1200 m
#' @returns "YES" or "?" for given SF point
#'
#' @usage check_atpol_square(centroid, raster, distance)
#' @export
#' @references
#'     \insertAllCited{}
#'
check_atpol_square <- function(centroid, raster, distance = 1200) {
  if (!grepl("ID\\[\"EPSG\",2180\\]\\]",terra::crs({{raster}}))) {
    warning("Raster has to be in EPSG:2180 projection")
    warning("Try terra::project(raster, \"EPSG:2180\")")
    stop()
  }
  c <- sf::st_buffer({{centroid}}, dist = {{distance}})
  mm <- terra::minmax(terra::crop({{raster}}, terra::vect(c)))
  y <- ifelse(mm[1] == 0 & mm[2] == 0, "YES", "?")
  return(y)
}


R/plot_points_on_atpol.R
# .myCols defines a colors used for classified raster colorization
.myCols <- c("#29AB08",
             "#4CBF00",
             "#7ACC00",
             "#ADD900",
             "#FFFF00",
             "#FFDD00",
             "#FFB700",
             "#FF9000",
             "#FF6A00",
             "#FF4400",
             "#E23924",
             "#FF1D00",
             "#F70000",
             "#D00000",
             "#AA0000")

#' plotPoitsOnAtpol() plots the observations on ATPOL 10km x 10km grid
#'
#' @importFrom grDevices dev.off png svg
#' @importFrom terra plot rast
#' @importFrom sf st_bbox
#' @importFrom graphics axis par
#' @param myData SimpleFeature data frame with point geometry, usually centroid of ATPOL grid square
#' @param outputType image output type, either "svg" or "png"; if not specified a standard output device is used (screen)
#' @param filename name of the output file
#' @param main image title, usually a species name
#' @param colors vector of colors to be used as a background, default internal .myCols. If FALSE, black and white plot has to be expected
#' @param water logical, if water layer has to be included, default TRUE
#' @param cex size of the points, default 0.9
#' @param col color of the points, default black
#' @param pch shape of the point, default 16 - filled dot
#' @param grid10k logical, default FALSE. If the small grid has to be printed out
#' @return choreograph map of species distribution in Poland.
#' @export
#' @usage plotPoitsOnAtpol(myData, outputType, filename, main, colors,
#' water, cex, col, pch, grid10k)

plotPoitsOnAtpol <- function(myData = "", outputType = "", filename = "", main = "",
                             colors = .myCols, water = TRUE,
                             cex = 0.9, col = "black", pch = 16, grid10k = FALSE) {
  oldpar <- par(no.readonly = TRUE)
  on.exit(par(oldpar))
  if(nzchar({{filename}}) == TRUE) {
    if(outputType == "svg") {
      grDevices::svg(file=paste0({{filename}}, ".svg"),width=12,height=12)
    } else if(outputType == "png"){
      grDevices::png(file=paste0({{filename}}, ".png"), width=1600, height=1600, res=240)
    }
    else {
      warning("Please provide an outputType (svg or png).")
    }
  }
  ### generate background
  .bbox <- sf::st_bbox(atpol10k())
  par(pty = "s")
  # a100k <- atpol100k()
  a100k <- setNames(aggregate(.atpol10k_full()[,"geometry"], by = list(substr(.atpol10k_full()$Name, 1, 2)), mean), c("Name", "geometry"))
  cr <- terra::rast(system.file("extdata/cr.tif", package = "atpolR"))
  if(!isFALSE({{colors}})) {
    terra::plot(cr, type="classes",
                col = colors,
                #        col = terrain.colors(12, alpha = 1, rev = FALSE),
                legend = FALSE,
                xlim = c(.bbox[1], .bbox[3]),
                ylim = c(.bbox[2], .bbox[4]),
                axes = FALSE,
                main = {{main}})

    if(isTRUE({{water}})) {
      # adding water layer
      w <- terra::rast(system.file("extdata/water.tif", package = "atpolR"))
      terra::plot(w, add = TRUE, col = "#4476C3", legend = FALSE, axes = FALSE)
    }
  } else {
    # although it doesn't make sense to plot white raster,
    # it's to have the same plot dimension
    terra::plot(cr, type="classes",
                col = "white",
                legend = FALSE,
                xlim = c(.bbox[1], .bbox[3]),
                ylim = c(.bbox[2], .bbox[4]),
                axes = FALSE,
                main = {{main}})
  }
  terra::plot(a100k$geometry, lwd = 0.3, add= TRUE)

  d <- list()
  for (letter in LETTERS[1:7]) {
    a <- subset(a100k, a100k$Name == paste0("D", letter)) |> sf::st_bbox()
    c <- a[4] - (a[4] - a[2])/2
    d <- append(d, c)
  }
  axis(2, at = c(d[1:7]), labels = c("A", "B", "C", "D", "E", "F", "G"), las = 1, lwd = 0, lwd.ticks = 0, line = -0.5, cex = 1.1)
  # axis(4, at = c(d[1:7]), labels = c("A", "B", "C", "D", "E", "F", "G"), las = 1, lwd = 0, lwd.ticks = 0, line = -0.5, cex = 1.1)

  d <- list()
  for (letter in LETTERS[1:7]) {
    a <- subset(a100k, a100k$Name == paste0(letter, "D")) |> sf::st_bbox()
    c <- a[3] - (a[3] - a[1])/2
    d <- append(d, c)
  }
  # axis(1, at = c(d[1:7]), labels = c("A", "B", "C", "D", "E", "F", "G"), las = 1, lwd = 0, lwd.ticks = 0, line = -0.5, cex = 1.1)
  axis(3, at = c(d[1:7]), labels = c("A", "B", "C", "D", "E", "F", "G"), las = 1, lwd = 0, lwd.ticks = 0, line = -0.5, cex = 1.1)

  if(isTRUE({{grid10k}})) {
    gr <- atpol10k()
    gr |>
      sf::st_geometry() |>
      sf::st_cast(to = "LINESTRING") |>
      terra::plot(lwd = 0.2, col = "grey30", add = TRUE)
  }

  if(isFALSE({{colors}}) & isFALSE({{water}})) {
    suppressMessages(sf::st_read(system.file("extdata/waters.gpkg", package = "atpolR"), layer = "rivers")) |>
      sf::st_geometry() |>
      terra::plot(add = TRUE, col = "blue")

    suppressMessages(sf::st_read(system.file("extdata/waters.gpkg", package = "atpolR"), layer = "water")) |>
      sf::st_geometry() |>
      terra::plot(col = "blue", border = "blue", add = TRUE)
  }

  terra::plot(boundaryPL(), col = "darkred", add = TRUE)

  ### adding data
  if(inherits({{myData}}, "sf") || inherits({{myData}}, "sfc")) {
    terra::plot({{myData}}, pch = {{pch}},
                cex = ifelse(outputType == "svg", 1.5, {{cex}}),
                add = TRUE, col = {{col}}, lty = 1)

  }

  if(nzchar({{filename}}) == TRUE && nzchar({{outputType}}) == TRUE) {
    grDevices::dev.off()
  }
}
